monix-reactive/jvm/src/main/scala/monix/reactive/compress/internal/operators/GunzipOperator.scala

```diff
@@ -0,0 +1,334 @@
+/*
+ * Copyright (c) 2014-2020 by The Monix Project Developers.
+ * See the project homepage at: https://monix.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package monix.reactive.compress.internal.operators
+
+import java.util.zip.{CRC32, DataFormatException, Inflater}
+import java.{util => ju}
+
+import monix.execution.Ack
+import monix.execution.Ack.{Continue, Stop}
+import monix.reactive.Observable.Operator
+import monix.reactive.compress.internal.operators.Gunzipper._
+import monix.reactive.compress.{
+  gzipCompressionMethod,
+  gzipFlag,
+  gzipMagicFirstByte,
+  gzipMagicSecondByte,
+  CompressionException
+}
+import monix.reactive.observers.Subscriber
+
+import scala.annotation.tailrec
+import scala.concurrent.Future
+import scala.util.control.NonFatal
+
+private[compress] final class GunzipOperator(bufferSize: Int) extends Operator[Array[Byte], Array[Byte]] {
+
+  def apply(out: Subscriber[Array[Byte]]): Subscriber[Array[Byte]] =
+    new Subscriber[Array[Byte]] {
+      implicit val scheduler = out.scheduler
+
+      private[this] var isDone = false
+      private[this] var ack: Future[Ack] = _
+      private[this] val gunzipper = new Gunzipper(bufferSize)
+
+      def onNext(elem: Array[Byte]): Future[Ack] = {
+        if (isDone) {
+          Stop
+        } else {
+          try {
+            val result = gunzipper.onChunk(elem)
+            // signaling downstream
+            ack = out.onNext(result)
+            ack
+          } catch {
+            case e if NonFatal(e) =>
+              onError(e)
+              Stop
+          }
+        }
+      }
+
+      def onError(ex: Throwable): Unit =
+        if (!isDone) {
+          isDone = true
+          gunzipper.close()
+          out.onError(ex)
+        }
+
+      def onComplete(): Unit =
+        if (!isDone) {
+          isDone = true
+          if (ack == null) ack = Continue
+          ack.syncOnContinue {
+            try {
+              out.onNext(gunzipper.finish())
```

Comment:
I think it should be:

```scala
      def onComplete(): Unit =
        if (!isDone) {
          isDone = true
          if (ack == null) ack = Continue

          ack.syncOnComplete {
            case Success(Continue) =>
              // Protect against contract violations - we are only allowed to
              // call onError if no other terminal event has been called.
              var streamErrors = true
              try {
                val lastArray = gunzipper.finish()
                streamErrors = false
                out.onNext(lastArray)
                out.onComplete()
              } catch {
                case NonFatal(e) if streamErrors =>
                  out.onError(e)
              } finally {
                gunzipper.close()
              }

            case _ => gunzipper.close()
          }
        }
```

For the following reasons:
1. We need to call `gunzipper.close()` even if downstream returned Stop / failed Future
2. If there is an error on `gunzipper.finish()` and we handle it by calling `out.onError`, we don't want to call `onComplete` as well
3. If there is an exception in `onNext` (protocol breakage) then we don't call onError in case it already happened downstream

Reply:
Yes, absolutely. I've just finished your recent talk about monix-reactive internals, so now it is all clear.